frontend/src/lib/stores/sns.store.ts

```
@@ -216,6 +221,42 @@ const initSnsQueryStore = (): SnsQueryStore => {
         ),
       }));
     },
+
+    /**
+     * Updates only the derived state of a sale.
+     *
+     * @param {Object} params
+     * @param {QuerySnsSwapState} params.swapData new swap data.
+     * @param {string} params.rootCanisterId canister id in text format.
+     */
+    updateLifecycle({
+      lifecycle,
+      rootCanisterId,
+    }: {
+      lifecycle: SnsSwapLifecycle;
+      rootCanisterId: string;
+    }) {
+      update((data: SnsQueryStoreData) => ({
+        metadata: data?.metadata ?? [],
+        swaps: (data?.swaps ?? []).map((swapData): QuerySnsSwapState => {
+          if (swapData.rootCanisterId === rootCanisterId) {
+            const swap = fromNullable(swapData.swap);
+            if (nonNullish(swap)) {
```

if there is a lifecycle and no swap then adding the lifecycle to the store is silently ignored, is that expected? there is no risk of race condition?

---

If there is no swap, nothing can be changed.

This is more a type issue. `swap: [] | [Swap]`. It's theoretically optional, but it's not really optional.

Ideally, and that's my plan for the future. I'd rather store each thing separately and put them together in a derived store.

But I don't want to do this here.

---

What I mean is that in

```
await Promise.all([
      loadSnsTotalCommitment({ rootCanisterId }),
     loadSnsLifecycle({ rootCanisterId }),
     loadCommitment(rootCanisterId),
    ]);
```

`loadSnsLifecycle` can finish and try to update store before `loadCommitment` does. In such case can it be a race condition? Like for example if the user refresh the project detail page?

---

No, this doesn't load on init of the page. Only after participation. And we reload only the data that we know might change after a participation:

* User commitment
* Total project commitment
* Project lifecycle.

There is already a project in store when this is called.

This is not about loading the whole project.

---

Good then, thanks for the feedback.